src/app/clusters/user-label-server/tests/TestUserLabelCluster.cpp

```diff
@@ -0,0 +1,186 @@
+#include <app/clusters/user-label-server/user-label-cluster.h>
+
+#include <app/clusters/testing/AttributeTesting.h>
+#include <app/server-cluster/AttributeListBuilder.h>
+#include <app/server-cluster/DefaultServerCluster.h>
+#include <app/server-cluster/testing/TestServerClusterContext.h>
+#include <clusters/UserLabel/Attributes.h>
+#include <clusters/UserLabel/Metadata.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::UserLabel;
+using namespace chip::app::Clusters::UserLabel::Attributes;
+
+namespace {
+
+struct TestUserLabelCluster : public ::testing::Test
+{
+    static void SetUpTestSuite() { ASSERT_EQ(Platform::MemoryInit(), CHIP_NO_ERROR); }
+    static void TearDownTestSuite() { Platform::MemoryShutdown(); }
+};
+
+constexpr size_t kUserLabelFixedClusterCount = 2;
+
+class UserLabelClusterTest
+{
+    UserLabelCluster userLabel;
+
+public:
+    template <typename... Args>
+    UserLabelClusterTest(Args &&... args) : userLabel(std::forward<Args>(args)...)
+    {}
+
+    template <typename F>
+    void Check(F check)
+    {
+        check(userLabel);
+    }
+};
+
+class ReadAttribute
+{
+    UserLabelCluster & mCluster;
+    DataModel::ActionReturnStatus mStatus;
+
+public:
+    ReadAttribute(UserLabelCluster & cluster) : mCluster(cluster), mStatus(CHIP_NO_ERROR) {}
+
+    DataModel::ActionReturnStatus GetStatus() const { return mStatus; }
+
+    void operator()(DataModel::ReadAttributeRequest & request)
+    {
+        Platform::ScopedMemoryBufferWithSize<uint8_t> buffer;
+        ASSERT_NE(buffer.Alloc(1024).Get(), nullptr);
+
+        AttributeReportIBs::Builder attributeReportIBsBuilder;
+        TLV::TLVWriter reportWriter;
+        reportWriter.Init(buffer.Get(), buffer.AllocatedSize());
+        ASSERT_EQ(attributeReportIBsBuilder.Init(&reportWriter), CHIP_NO_ERROR);
+
+        AttributeValueEncoder encoder(attributeReportIBsBuilder, Access::SubjectDescriptor{}, request.path, 0 /* dataVersion */);
+
+        mStatus = mCluster.ReadAttribute(request, encoder);
+    }
+};
+
+class WriteAttribute
+{
+    UserLabelCluster & mCluster;
+    DataModel::ActionReturnStatus mStatus;
+
+public:
+    WriteAttribute(UserLabelCluster & cluster) : mCluster(cluster), mStatus(CHIP_NO_ERROR) {}
+
+    DataModel::ActionReturnStatus GetStatus() const { return mStatus; }
+
+    void operator()(DataModel::WriteAttributeRequest & request)
+    {
+        Platform::ScopedMemoryBufferWithSize<uint8_t> buffer;
+        ASSERT_NE(buffer.Alloc(1024).Get(), nullptr);
+
+        TLV::TLVReader reportReader;
+        reportReader.Init(buffer.Get(), buffer.AllocatedSize());
+
+        AttributeValueDecoder decoder(reportReader, Access::SubjectDescriptor{});
+
+        mStatus = mCluster.WriteAttribute(request, decoder);
+    }
+};
+
+} // namespace
+
+TEST_F(TestUserLabelCluster, AttributeTest)
+{
+    for (EndpointId endpoint = 0; endpoint < kUserLabelFixedClusterCount; ++endpoint)
```

Comment:
what do these loops acomplish? Why not have some kTestEndpointId = 100 or something and test that. I doubt this is an edge case in any way and seems to make the test more complex.

Reply:
In that case we do not even need the UserLabelTes wrapper or anything just a cluster